### ISSUE TITLE
parallel/grid computing

### ISSUE THREAD

**Issue by nicholasjhorton** [hglanz/DS-Blog-Ex#61]
Thoughts:
- a brief mention of parallel computation to do embarassingly parallel things (simulation study?)
- a shoutout to Spark and sparklyr to do the same thing?

Or swap this out for a different topic (reprex?  testthat? next steps for causal inference?).  This one feels like it might be too advanced for our intended audience.

**Comment by hglanz** [comment]
@hardin47 I'm getting the following error when I try to build the staging site, from parallel.Rmd

Error in summary.connection(connection) : invalid connection

Any ideas?

**Comment by hardin47** [comment]
sorry about that.  i was just messing around with the parallel code yesterday in light of upcoming blog.  i put `eval=FALSE` and re-pushed, so building should work now.

meanwhile, i was actually trying to make the exact point that there is an invalid connection at that line of code!  the point is about connecting R to the idea of a cluster.  Here is what the output of that code chunk should look like:

![image](https://user-images.githubusercontent.com/7210585/59764744-40485800-9251-11e9-9cee-f1c3e0346898.png)


so i think i want to be able to put in some kind of "don't break even if there is an error" option in the code chunk preamble.  is that possible?

thanks!

**Comment by nicholasjhorton** [comment]
You can add "error = TRUE" as a code chunk option and it will happily continue.

**Comment by nicholasjhorton** [comment]
consider adding a shoutout to the RStudio cheatsheet contribution on parallel computing?   https://www.rstudio.com/resources/cheatsheets/

**Comment by hardin47** [comment]
great, done, thanks!  i'm still working on the post, it isn't ready for review yet.

**Comment by hardin47** [comment]
ok, i finished the parallel computing post.  i'd love feedback from @hglanz and @nicholasjhorton (I know you are headed out of town, Hunter!).  i had some Java issues on my own end (you need Java 8 to do `sparklyr`).  i hope that it doesn't impact the larger build.

**Comment by nicholasjhorton** [comment]
I *finally* got myself set up to be able to run this file (Java was particularly frustrating for me). It's looking good.  Thanks for your work on this front.

Some additional requests (in decreasing order of importance from my vantage point):

- [ ] clip out the "Master-slave Paradigm" from the first figure
- [ ] add a paragraph early on discussing the importance of parallel computing that introduces the ideas of distributed computing, cluster computing, and grid computing under a framework of high performance computing
- [ ] consider adding a pointer to https://en.wikipedia.org/wiki/Randomized_algorithm
- [ ] consider adding a pointer to https://en.wikipedia.org/wiki/Grid_computing
- [ ] close with a paragraph about how students need sufficient background in CS to be able to take courses on the principled used of high performance computing
- [ ] consider adding a pointer to this blog: https://livefreeordichotomize.com/2019/06/04/using_awk_and_r_to_parse_25tb/

**Comment by hardin47** [comment]
i'm working on this right now, but i have a few questions on adding @nicholasjhorton  pointers.

why: https://en.wikipedia.org/wiki/Randomized_algorithm ?  it is because the idea of randomized_algorithm is also embarrassingly parallel?

why:  https://livefreeordichotomize.com/2019/06/04/using_awk_and_r_to_parse_25tb/ ?  is it because he provides an example of spark computing in the wild?

**Comment by hardin47** [comment]
thanks for the edits @nicholasjhorton !  i pushed my edits here:  https://github.com/hglanz/DS-Blog-Ex/commit/d517aae888e1053d642b48833e2510d10da1e597

i didn't do anything with the one comment on randomized_algorithm.  i'm not sure where to put it.

i'm ready to close this issue.

**Comment by nicholasjhorton** [comment]
On Jul 5, 2019, at 11:32 AM, Jo Hardin <notifications@github.com> wrote:

why: https://en.wikipedia.org/wiki/Randomized_algorithm ? it is because the idea of randomized_algorithm is also embarrassingly parallel?

MY RESPONSE: A CS member of the Data Science Education roundtable accused statisticians of not knowing about randomized algorithms.  I'd like to weave the importance of algorithmic thinking into the entry.  Perhaps something as simple as  "Data science methods increasingly use [randomized algorithms](https://en.wikipedia.org/wiki/Randomized_algorithm).

why:  https://livefreeordichotomize.com/2019/06/04/using_awk_and_r_to_parse_25tb/ ? is it because he provides an example of spark computing in the wild?

MY RESPONSE: Yes, with some really neat insights.  I like this because it gives a sense of the challenges when actually moves to scale.  (Note that our examples for this entry are really toy examples.  I'm fine with this but would like to mention a more sophisticated mention.)


Nicholas Horton 
Beitzel Professor of Technology and Society (Statistics and Data Science)
Amherst College

**Comment by hardin47** [comment]
Ok, I added it here, but I suppose I don't really understand it, so I'm not sure I added it correctly (the CS member was correct!).  @nicholasjhorton feel free to fix if it doesn't read correctly.

In this blog, we'll perform tasks that are **embarassingly parallel** which means there is no dependency or communication between the parallel tasks.  Again, parallel computing can be powerful in ways that link computational tasks in complicated ways.  But we believe that as a *first* pass at teaching parallel computing, we should teach the parallel structure before bringing in dependence across the parallel tasks.  Examples of [embarrassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel) algorithms include: Monte Carlo analysis, bootstrapping, growing trees for Random Forests, `group_by` analyses, and cross-validation.  Additionally, data science methods increasingly use [randomized algorithms](https://en.wikipedia.org/wiki/Randomized_algorithm) which can often be written in parallel.

**Comment by nicholasjhorton** [comment]
That's excellent: thanks for incorporating my suggestion!

Nick


On Jul 5, 2019, at 5:03 PM, Jo Hardin <notifications@github.com> wrote:

Ok, I added it here, but I suppose I don't really understand it, so I'm not sure I added it correctly (the CS member was correct!). @nicholasjhorton feel free to fix if it doesn't read correctly.

In this blog, we'll perform tasks that are embarassingly parallel which means there is no dependency or communication between the parallel tasks. Again, parallel computing can be powerful in ways that link computational tasks in complicated ways. But we believe that as a first pass at teaching parallel computing, we should teach the parallel structure before bringing in dependence across the parallel tasks. Examples of embarrassingly parallel algorithms include: Monte Carlo analysis, bootstrapping, growing trees for Random Forests, group_by analyses, and cross-validation. Additionally, data science methods increasingly use randomized algorithms which can often be written in parallel.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


Nicholas Horton 
Beitzel Professor of Technology and Society (Statistics and Data Science)
Amherst College
